Appending to TSDB: Move crucial actions to defer

beorn7 opened the pull request:
With defer having less of a performance penalty, there is no reason
not to do those crucial operations via defer.

Context: With isolation in place, if we forget to Commit/Rollback, the
low watermark will get stuck forever.

The current code should not have any bugs, but moving to defer helps
to avoid future bugs.

This is also moving the `closeAppend` in the `Commit` implementation
itself to defer. If logging to the WAL fails, we would have missed the
`closeAppend`.

Related to #6909 .

roidelapluie commented:
I put my release shepherd veto on this pull request until we use go1.14, which we are not at this moment.

beorn7 commented:
Yes, sure.

For context: https://github.com/prometheus/golang-builder/pull/85 describes what is holding us back.

roidelapluie commented:
> Yes, sure.
> 
> For context: [prometheus/golang-builder#85](https://github.com/prometheus/golang-builder/pull/85) describes what is holding us back.

Yes, sorry, github does not send updated comments by email.

brian-brazil commented on scrape/scrape.go:
```diff
@@ -1258,67 +1259,71 @@ const (
-		app.Rollback()
-		return err
+	if err = sl.addReportSample(app, scrapeHealthMetricName, ts, health); err != nil {
+		return
```
It's a little weird to see returning without the err, could we make this explicit?

beorn7 replied:
While I avoid naked returns in general, in this particular case it is a legitimate pattern (inspecting and changing the returned error in `defer`). And In that case, I'd like to embrace the naked return completely and mark it clearly by, well, a naked return.

beorn7 replied:
Having said that, if we have a style convention to still put the names in the return statement, I'm happy to do so, but I'm not aware of any.

brian-brazil commented on tsdb/head.go:
```diff
@@ -1068,13 +1069,14 @@ func (a *headAppender) Commit() error {
 
 func (a *headAppender) Rollback() error {
-	a.head.metrics.activeAppenders.Dec()
+	defer a.head.metrics.activeAppenders.Dec()
```
Why defer rather than immediately?

beorn7 replied:
To be consistent with what we do in `Commit`. I also think it is clearer to do it in the end in this way. Makes debugging easier if for example one of the locks cannot get acquired.

roidelapluie commented on scrape/scrape.go:
```diff
@@ -1258,67 +1259,71 @@ const (
 	app := sl.appender()
+	defer func() {
+		if err != nil {
+			app.Rollback()
```
while we are at it should we use multierr ?

beorn7 replied:
I'd like to keep unrelated changes out of this PR.

beorn7 replied:
Or perhaps not completely unrelated, depending on how you see it.

But still, let's do one (non-trivial) thing per PR.

beorn7 commented:
I have rebased this on top of the release-2.17 branch and changed the base.

Ironically, this doesn't really help with the panic in #6977 but I think, now that we have seen panics getting recovered because they are in the HTTP serving path, we should be conservative and merge this into release-2.17.

I'll keep checking the querying path for similar requirements to move to defer.

roidelapluie commented:
Can we reach consensus regarding the explicit returns discussion? @brian-brazil @beorn7

roidelapluie commented:
> Ironically, this doesn't really help with the panic in #6977

Can we do something for that kind of panics?

beorn7 commented:
> Can we do something for that kind of panics?

Done in #6983 . (o:

brian-brazil commented:
> Can we reach consensus regarding the explicit returns discussion?

I'm not that pushed on it.

roidelapluie commented:
Thanks!
